### PR TITLE
WIP: MOI wrapper

### PR DESCRIPTION
**Pull request by Wikunia**
First state of MOI wrapper
- Needs more test cases 
- Fix todos like
   - TimeLimit for MIP in fpump
   - Compute objective i.e in `generate_real_nlp`

### CODE REVIEW

**Comment by ccoffrin** [comment]
Amazing work, thanks so much!!!

File: REQUIRE [hunk 1]
```diff
@@ -1,5 +1,6 @@
-julia 0.6
+julia 1.0
 
 MathProgBase 0.6 0.8-
-JuMP 0.18 0.19-
+MathOptInterface
+JuMP 0.19
```
Comment (odow) [inline REQUIRE]: You should put an upper bound on JuMP

File: src/Juniper.jl [hunk 2]
```diff
@@ -4,6 +4,16 @@ using MathProgBase
 using JuMP
```
Comment (odow) [inline src/Juniper.jl]: You probably want to use `import JuMP` and prefix all JuMP functions with `JuMP.`.
http://www.juliaopt.org/JuMP.jl/latest/style/#using-vs.-import-1

File: src/MOI_wrapper/results.jl [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	
+function MOI.get(model::Optimizer, ::MOI.TerminationStatus)
+	return model.inner.status
+end
+	
+function MOI.get(model::Optimizer, ::MOI.ObjectiveValue)
+    if model.inner.status == MOI.OPTIMIZE_NOT_CALLED
+	    @error "optimizer not called"
+	end
+	return model.inner.objval
```
Comment (odow) [inline src/MOI_wrapper/results.jl]: There is some weird indentation in this file

File: REQUIRE (hunk 1 above)
Comment (Wikunia) [reply REQUIRE]: set to 0.19 0.20-

File: src/Juniper.jl (hunk 2 above)
Comment (Wikunia) [reply src/Juniper.jl]: now `import JuMP` and using macros like `@variable,@constraint,@objective`

File: src/MOI_wrapper/results.jl (hunk 3 above)
Comment (Wikunia) [reply src/MOI_wrapper/results.jl]: not reproducible on my machine with Visual Studio Code will check again.

**Comment by Wikunia** [comment]
Regarding CodeCov: 
- Any idea why these lines are red but the one before and after are green?
https://codecov.io/gh/lanl-ansi/Juniper.jl/src/moi-wrapper/src/solver.jl#L10...24
- Shell we have test cases for `MOI.support` like these: 
https://codecov.io/gh/lanl-ansi/Juniper.jl/src/3c2634ed82d430871cef9a292926ee4fcc7cf21c/src/MOI_wrapper/constraints.jl#L8...10

**Comment by Wikunia** [comment]
Additionally the Readme and docs need to be updated as well.

**Comment by odow** [comment]
Coverage issue is: https://github.com/JuliaCI/Coverage.jl/issues/187

You'll probably hit https://github.com/JuliaWeb/MbedTLS.jl/issues/193 as well.

**Comment by kaarthiksundar** [comment]
@Wikunia Please list out the things that you think that remain as a part of this MOI PR, so that we can finish them one by one and merge this PR. Meanwhile, @ccoffrin said he will run this branch with the MINLPTests.

**Comment by Wikunia** [comment]
Do we have to free the model for Knitro with i.e. `Base.finalize`? or is it entirely the problem of Knitro if it's not working?
Do we want to try to add a time limit to Cbc in the MIP part of the fpump?
- This can probably be done similarly as setting `mu_init` for Ipopt 
Is everything tested that needs to be tested as codecov is not really helping.
Maybe getting a test case for this: 
https://codecov.io/gh/lanl-ansi/Juniper.jl/src/moi-wrapper/src/bb_strategies.jl#L147...149
Not sure why it doesn't get hit as the test cases didn't change.

**Comment by frapac** [comment]
@Wikunia What problem do you encounter with Knitro? Do you want me to investigate the issue?

**Comment by Wikunia** [comment]
@frapac I don't have a knitro license so am unable to test it. At LANL we had a problem of calling knitro as the nl solver instead of Ipopt as of license restrictions. The license wasn't freed after optimizing a subproblem. We ran `MPB.freemodel` for that reason now that function doesn't exist anymore. There might be changes in Knitro which solves that issue already. Maybe @kaarthiksundar can run a test and report back.

**Comment by kaarthiksundar** [comment]
@frapac I will get you an example

**Comment by frapac** [comment]
We add a support to Knitro's license manager in latest release of KNITRO.jl: 
https://github.com/JuliaOpt/KNITRO.jl/blob/master/src/kn_env.jl#L10

I think that should solve your issue. I can try to run Juniper with Knitro on my computer if that could be of any help. An example would help.

**Comment by Wikunia** [comment]
Handle `ALMOST_LOCALLY_SOLVED`:
Possibilities: 
Root node:
- handle it as locally solved 
- handle it as local infeasible and restart if restart is set
Normal tree node:
- Continue as locally solved + Warning
Integral Leaf node:
- Depends on user setting whether it gets accepted.
- Default: Gets accepted with warning

**Comment by Wikunia** [comment]
Missing before merging
- [ ] Handle `ALMOST_LOCALLY_SOLVED` @Wikunia 
- [ ] Check if Knitro is working @kaarthiksundar @frapac 
- [ ]  Check why travis errors on Julia v1.0 for Mac @kaarthiksundar 
- [ ]  Making a test run on MINLP problems @Wikunia @ccoffrin 

Currently Juniper v0.2.5 is better than this branch:
juniper-moi

**Comment by frapac** [comment]
I was able to test Knitro on the branch "moi-wrapper" of Juniper. 
In `test/runtests.jl`, line 48, I replaced `nl_solver` by: 
```
nl_solver=with_optimizer(KNITRO.Optimizer, outlev=0) 
```
Then, I ran the unit-tests. Everything is working fine, except: 
- I get some problems with the multiprocessing (I do not think it is related to Knitro); 
- I get an error in `test/parallel.jl`, line 360 (`infeasible relaxation`) 

This is an issue with the latest versions of KNITRO.jl and JuMP. When we solve a NLP problem through JuMP, and call `optimize!` twice in a row 
```
JuMP.optimize!(model)
JuMP.optimize!(model)
```
JuMP tries to redefine the NLP part of the problem, and that is not supported by KNITRO.jl. Refer to:
https://github.com/JuliaOpt/JuMP.jl/blob/master/src/optimizer_interface.jl#L105-#L108

Considering the TODO in the comments, I think we should patch JuMP to support incremental changes for NLP models.

**Comment by Wikunia** [comment]
Thanks @frapac 
> * I get some problems with the multiprocessing (I do not think it is related to Knitro);
Can it be the case that several licenses are needed then as problems gets solved in parallel?
I think this is the reason why we never tried it. Correct @ccoffrin ?
> This is an issue with the latest versions of KNITRO.jl and JuMP. When we solve a NLP problem through JuMP, and call `optimize!` twice in a row
> 
> ```
> JuMP.optimize!(model)
> JuMP.optimize!(model)
> ```
You mean restarting the root relaxation by simply changing the primal start values?

**Comment by kaarthiksundar** [comment]
@Wikunia In my understanding, I think it is because JuMP redefines the problem when optimize is called again, and Knitro.jl does not free the environment in case of incremental changes - https://github.com/JuliaOpt/KNITRO.jl/blob/da92f5a4b0510688b75ad92a314a3f0dd481ae6e/src/MOI_wrapper.jl#L213
This is my guess, I think @frapac will have a better answer.

**Comment by frapac** [comment]
@kaarthiksundar You are exactly right! I will update KNITRO.jl to free the model if `optimize!` is called several times. However, I think a long term solution would be support incremental changes directly inside JuMP to leverage solver warm starts!

**Comment by kaarthiksundar** [comment]
@frapac I agree with your long term solution. I think JuMP should not be rebuilding the model in case of incremental changes. But actually from what I understand, if a solver supports direct mode, it does not rebuild the model. Again, the only solver that supports direct mode is Gurobi (to the best of my knowledge - @odow). Furthermore, I do not know what needs to be done on a solver level to support this direct mode (may be @odow can comment).

**Comment by kaarthiksundar** [comment]
@Wikunia I ran the test in Julia v1.0 in my Mac and it runs fine. I think the issue is that the penultimate test takes too long and travis terminates the build if it has not received any response within a certain time limit. There are two easy fixes for this (1) switch on logging for that test or (2) turn off that test and replace it with a quicker one. Let me know which is easy. From the latest build, it seems like linux build also fails.

**Comment by odow** [comment]
> Again, the only solver that supports direct mode is Gurobi

All of the LQOI solvers (CLP, CPLEX, GLPK, Gurobi, Xpress) support direct mode. 

> I do not know what needs to be done on a solver level to support this direct mode 

Basically, the solver needs to support `add_variable` and `add_constraint` instead of the `copy_to` API. Cbc (for example), falls into the second category, and so can't be used in direct mode.

**Comment by Wikunia** [comment]
After implementing `ALMOST_LOCALLY_SOLVED` this is the updated result:
juniper-v0 2 5_vs_moi

MOI solves 5 instances less and 1 more than v0.2.5
- 3/5 time limit (>1h vs < 30 min in v0.2.5)
- one Iteration Limit in root node (gasnet_al3)
- one error in MIP which has nothing to do with this branch (deb6)

**Comment by ccoffrin** [comment]
These results LGTM.

**Comment by Wikunia** [comment]
Anyone an idea on the current bug in travis for the documentation?
Will you run it on the larger test set as well @ccoffrin ?
I think we can merge then and I'll try to narrow down where this moi version is slower the next weeks but we would have at least a working moi version.

**Comment by ccoffrin** [comment]
I'm working on running the large on MOI test presently.  Before merging we should make sure that at least most, and ideally all, of MINLPTests#moi pass.  Once that is done we should setup a travis build stage to check it.  Here is an example of how to set this up, https://github.com/JuliaOpt/AmplNLWriter.jl/blob/master/.travis.yml

**Comment by ccoffrin** [comment]
@Wikunia, @kaarthiksundar 

I had to add a basic implementaiton of

**Comment by ccoffrin** [comment]
@Wikunia, @kaarthiksundar 

I had to add a basic implementation of `DualStatus` to make this work in PowerModels (see, https://github.com/lanl-ansi/Juniper.jl/blob/moi-wrapper/src/MOI_wrapper/results.jl#L15).  Please check the semantics are correct.

Also I think we should add some unit tests for the three status parameters, termination, primal, and dual.  Once that is done, we can merge this and tag Juniper v0.3!

**Comment by ccoffrin** [comment]
Also Juniper MOI works with MIQCQP, which is a very nice feature over the MPB version.  Thanks for that!!!

**Comment by Wikunia** [comment]
Will we have problems regarding the documentation creation based on the travis failure?

**Comment by ccoffrin** [comment]
Lets fix the docs builder once the code is on master.  I don't know how to configure it for a branch.

**Pull request merged**